### PR TITLE
Test case for task history cache corruption

### PR DESCRIPTION
**Pull request by wolfs**
When Gradle is aborted after removing the obsolete snapshots but before persisting the new execution snapshot then the old task history cannot be loaded any more and we may experience problems as in https://github.com/gradle/gradle/issues/2827. We already fixed the underlying problem in #2870. This PR add an integration test exposing the problem.

### CODE REVIEW

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            requireGradleDistribution()
+        }
+
+        when:
+        succeeds("createFiles")
+        succeeds("clean")
+        fails("createFiles", "-PkillMe=true", "--max-workers=${numberOfTasks}")
+
+        then:
+        file('build/output10').allDescendants().size() == numberOfFiles
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: Is this check necessary? Couldnt a loaded agent fail to create all of these files before the daemon is killed?

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    void createFiles() {
+        workerExecutor.submit(CreateFiles) {
+            isolationMode = IsolationMode.NONE
+            params(outputDir)
+        }
+        if (project.findProperty("killMe")) {
+            new Thread({
+                Thread.sleep(${millisecondsToKill})
+                System.exit(1)
+            }).start()
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: Should we try to have a single kill thread/task?

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        then:
+        file('build/output10').allDescendants().size() == numberOfFiles
+
+        expect:
+        succeeds "createFiles"
+    }
+
+    private void setupTestProject(int numberOfFiles, int numberOfInputProperties, int numberOfTasks, int millisecondsToKill) {
+        buildFile << """
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: For our future selves, we may want a comment here describing how this project works and why its doing what it's doing.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: This seems to be one of the first tasks to be executed. I want the build to be killed when it is more or less halfway through. How about I add a check that the total number of files in the build directory is between `#tasks * #filesPerTask * (0.2 - 0.8)`? so 20% to 80% of all the files have been created?

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: I wouldn't add the extra complexity. Like this we can kill the build some time after the first task finished writing its files. WDYT?

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: I made the change I proposed.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy]: I added a comment.

**Comment by wolfs** [comment]
@big-guy I addressed your comments, could you have another look?

**Pull request merged**